File: python-logger/zoplog_config.py
#!/usr/bin/env python3
"""
Centralized configuration loader for ZopLog
Reads from /etc/zoplog/database.conf with fallbacks
"""
import os
import json
import configparser
from typing import Dict, Any

DEFAULT_CONFIG_PATHS = [
    "/etc/zoplog/database.conf",
]

def load_database_config() -> Dict[str, Any]:
    """Load database configuration from centralized location"""
    config = {
        "host": os.getenv("ZOPLOG_DB_HOST", "localhost"),
        "user": os.getenv("ZOPLOG_DB_USER", "zoplog_db"),
        "password": os.getenv("ZOPLOG_DB_PASS", ""),
        "database": os.getenv("ZOPLOG_DB_NAME", "logs_db"),
        "port": int(os.getenv("ZOPLOG_DB_PORT", "3306")),
    }
    
    # Try to read from config files
    for config_path in DEFAULT_CONFIG_PATHS:
        if os.path.exists(config_path) and os.path.isfile(config_path):
            try:
                if config_path.endswith('.conf'):
                    # INI-style config file
                    parser = configparser.ConfigParser()
                    parser.read(config_path)
                    if 'database' in parser:
                        db_section = parser['database']
                        config.update({
                            "host": db_section.get('host', config['host']),
                            "user": db_section.get('user', config['user']),
                            "password": db_section.get('password', config['password']),
                            "database": db_section.get('name', config['database']),
                            "port": int(db_section.get('port', config['port'])),
                        })
                        break
                else:
                    # Legacy key=value format
                    with open(config_path, 'r') as f:
                        for line in f:
                            line = line.strip()
                            if not line or line.startswith('#'):
                                continue
                            if '=' in line:
                                key, value = line.split('=', 1)
                                key = key.strip().upper()
                                value = value.strip()
                                if key in ('DB_HOST', 'HOST'):
                                    config['host'] = value
                                elif key in ('DB_USER', 'USER'):
                                    config['user'] = value
                                elif key in ('DB_PASS', 'PASSWORD', 'PASS'):
                                    config['password'] = value
                                elif key in ('DB_NAME', 'DATABASE', 'NAME'):
                                    config['database'] = value
                                elif key in ('DB_PORT', 'PORT'):
                                    config['port'] = int(value)
                    break
            except Exception as e:
                print(f"Warning: Could not read config from {config_path}: {e}")
                continue
    
    # Final fallback for password if still empty
    if not config['password']:
        config['password'] = 'uxaz0F0XMDAG7X1pVRF8yGCTo5sT7dnYNAVwZukags4='
    
    return config

def load_settings_config() -> Dict[str, Any]:
    """Load system settings from centralized config"""
    config_paths = [
        "/etc/zoplog/zoplog.conf",      # Centralized location
    ]
    
    # Default settings with eth0 as preferred interface for internet monitoring
    defaults = {
        "monitor_interface": "eth0",    # WAN interface for better internet traffic capture
        "firewall_interface": "eth0",   # Apply firewall to internet-facing interface
        "capture_mode": "promiscuous",
        "log_level": "INFO",
        "block_mode": "immediate",
        "log_blocked": True,
        "update_interval": 30,
        "max_log_entries": 10000
    }
    
    for config_path in config_paths:
        try:
            if os.path.exists(config_path):
                if config_path.endswith('.conf'):
                    # INI format - new centralized config
                    import configparser
                    parser = configparser.ConfigParser()
                    parser.read(config_path)
                    
                    config = defaults.copy()
                    if parser.has_section('monitoring'):
                        monitoring = parser['monitoring']
                        config['monitor_interface'] = monitoring.get('interface', config['monitor_interface'])
                        config['capture_mode'] = monitoring.get('capture_mode', config['capture_mode'])
                        config['log_level'] = monitoring.get('log_level', config['log_level'])
                    
                    if parser.has_section('firewall'):
                        firewall = parser['firewall']
                        config['firewall_interface'] = firewall.get('apply_to_interface', config['firewall_interface'])
                        config['block_mode'] = firewall.get('block_mode', config['block_mode'])
                        config['log_blocked'] = firewall.getboolean('log_blocked', config['log_blocked'])
                    
                    if parser.has_section('system'):
                        system = parser['system']
                        config['update_interval'] = system.getint('update_interval', config['update_interval'])
                        config['max_log_entries'] = system.getint('max_log_entries', config['max_log_entries'])
                    
                    return config
                else:
                    # JSON format - legacy
                    with open(config_path, 'r') as f:
                        settings = json.load(f)
                        return {**defaults, **settings}
        except Exception as e:
            print(f"Warning: Could not read settings from {config_path}: {e}")
            continue
    
    return defaults

# Backwards compatibility
DB_CONFIG = load_database_config()
DEFAULT_MONITOR_INTERFACE = load_settings_config().get("monitor_interface", "eth0")

File: web-interface/zoplog_config.php
<?php
/**
 * Centralized ZopLog Database Configuration
 * Reads from /etc/zoplog/database.conf with fallbacks
 */

function load_zoplog_database_config() {
    // Default configuration
    $config = [
        'host' => getenv('ZOPLOG_DB_HOST') ?: 'localhost',
        'user' => getenv('ZOPLOG_DB_USER') ?: 'root', 
        'password' => getenv('ZOPLOG_DB_PASS') ?: '',
        'database' => getenv('ZOPLOG_DB_NAME') ?: 'logs_db',
        'port' => (int)(getenv('ZOPLOG_DB_PORT') ?: 3306),
    ];
    
    // Try config file locations in order
    $config_paths = [
        '/etc/zoplog/database.conf',
    ];
    
    foreach ($config_paths as $path) {
        if (!file_exists($path)) continue;
        
        try {
            if (pathinfo($path, PATHINFO_EXTENSION) === 'conf') {
                // INI-style config file
                $ini = parse_ini_file($path, true);
                if (isset($ini['database'])) {
                    $db_section = $ini['database'];
                    $config['host'] = $db_section['host'] ?? $config['host'];
                    $config['user'] = $db_section['user'] ?? $config['user'];
                    $config['password'] = $db_section['password'] ?? $config['password'];
                    $config['database'] = $db_section['name'] ?? $config['database'];
                    $config['port'] = (int)($db_section['port'] ?? $config['port']);
                    break;
                }
            } else {
                // Legacy key=value format
                $lines = file($path, FILE_IGNORE_NEW_LINES | FILE_SKIP_EMPTY_LINES);
                if ($lines !== false) {
                    foreach ($lines as $line) {
                        $line = trim($line);
                        if ($line === '' || $line[0] === '#') continue;
                        
                        $parts = explode('=', $line, 2);
                        if (count($parts) === 2) {
                            $key = strtoupper(trim($parts[0]));
                            $value = trim($parts[1]);
                            
                            switch ($key) {
                                case 'DB_HOST':
                                case 'HOST':
                                    $config['host'] = $value;
                                    break;
                                case 'DB_USER':
                                case 'USER':
                                    $config['user'] = $value;
                                    break;
                                case 'DB_PASS':
                                case 'PASSWORD':
                                case 'PASS':
                                    $config['password'] = $value;
                                    break;
                                case 'DB_NAME':
                                case 'DATABASE':
                                case 'NAME':
                                    $config['database'] = $value;
                                    break;
                                case 'DB_PORT':
                                case 'PORT':
                                    $config['port'] = (int)$value;
                                    break;
                            }
                        }
                    }
                    break;
                }
            }
        } catch (Exception $e) {
            error_log("Warning: Could not read ZopLog config from $path: " . $e->getMessage());
            continue;
        }
    }
    
    // Final fallback for password
    if (empty($config['password'])) {
        $config['password'] = 'uxaz0F0XMDAG7X1pVRF8yGCTo5sT7dnYNAVwZukags4=';
    }
    
    return $config;
}

// Load configuration and create connection
$db_config = load_zoplog_database_config();

$mysqli = @new mysqli(
    $db_config['host'],
    $db_config['user'], 
    $db_config['password'],
    $db_config['database'],
    $db_config['port']
);

if ($mysqli->connect_errno) {
    http_response_code(500);
    die('Database connection failed: ' . htmlspecialchars($mysqli->connect_error));
}

$mysqli->set_charset('utf8mb4');

// Export for backwards compatibility
$DB_HOST = $db_config['host'];
$DB_USER = $db_config['user'];
$DB_PASS = $db_config['password'];
$DB_NAME = $db_config['database'];
$DB_PORT = $db_config['port'];
?>
